src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,261 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            {
+                if (!inputType.IsErrorType() && !hasErrors)
+                {
+                    var errorCode = node.IsKind(SyntaxKind.ListPatternClause)
+                        ? ErrorCode.ERR_UnsupportedTypeForListPattern
+                        : ErrorCode.ERR_UnsupportedTypeForLengthPattern;
+                    Error(diagnostics, errorCode, node, inputType.ToDisplayString());
+                }
+                hasErrors = true;
+            }
+        }
+
+        private BoundListPatternClause BindListPatternClause(
+            PropertyPatternClauseSyntax node,
+            TypeSymbol inputType,
+            bool permitDesignations,
+            ref PropertySymbol? lengthProperty,
+            ref bool hasErrors,
+            BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause));
+
+            if (lengthProperty is null)
+            {
+                LookupLengthOrCountProperty(node, inputType, out lengthProperty, ref hasErrors, diagnostics);
+            }
+
+            TypeSymbol elementType;
+            IndexerArgumentInfo? info = null;
+            if (inputType.IsSZArray())
+            {
+                elementType = ((ArrayTypeSymbol)inputType).ElementType;
+            }
+            else if (TryFindIndexIndexerPatternForListPattern(node, inputType, out info, diagnostics))
+            {
+                elementType = info.Symbol.GetTypeOrReturnType().Type;
+            }
+            else
+            {
+                // PROTOTYPE(list-patterns) Cover any other type that we intend to cover, multi-dimensional arrays?
+                if (!hasErrors && !inputType.IsErrorType())
+                    Error(diagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+                hasErrors = true;
+                elementType = CreateErrorType();
+            }
+
+            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
+            return new BoundListPatternClause(node, elementType, subpatterns, sawSlice, info, hasErrors);
+        }
+
+        private bool TryFindIndexIndexerPatternForListPattern(
+            SyntaxNode syntax, TypeSymbol receiverType,
+            [NotNullWhen(true)] out IndexerArgumentInfo? info,
+            BindingDiagnosticBag diagnostics)
+        {
+            using var _ = LookupResult.GetInstance(out var lookupResult);
+            return PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out info, Compilation.GetWellKnownType(WellKnownType.System_Index), diagnostics) ||
+                   PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out info, Compilation.GetSpecialType(SpecialType.System_Int32), diagnostics);
+        }
+
+        private bool TryFindRangeIndexerPatternForListPattern(
```

Comment:
nit: Current name suggests this only for pattern-based range indexer.  Consider renaming `TryFindSlicingInfo`.
If that sounds okay, then maybe `TryFindIndexIndexerPatternForListPattern` -> `TryFindIndexingInfo`

nit: Also consider making those two methods local functions (only used in one method).